File: internal/schedule/schedule_test.go
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

package schedule_test

import (
	"context"
	"net/http"
	"os"
	"path/filepath"
	"testing"

	"github.com/steadybit/cli/internal/jsyaml"
	"github.com/steadybit/cli/internal/platformtest"
	"github.com/steadybit/cli/internal/schedule"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var ctx = context.Background()

const id = "01951394-727f-76a0-8675-c7519ebd0ff5"

var saved = map[string]any{
	"id": id, "experimentKey": "ADM-1", "cron": "0 0 9 ? * MON-FRI", "timezone": "Europe/Berlin", "enabled": true,
	"allowParallel": false, "editedBy": map[string]any{"username": "jane"}, "lastUpdated": "2026-09-01T10:00:00Z", "nextExecution": "2026-09-02T07:00:00Z",
}

func TestListFiltersWithRepeatedParameters(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/schedules/v2", platformtest.Reply{JSON: []any{saved}})

	out, err := platformtest.Stdout(t, func() error {
		return schedule.List(ctx, p.Client, schedule.ListOptions{Teams: []string{"ADM", "OPS"}, Experiments: []string{"ADM-1"}})
	})

	require.NoError(t, err)
	assert.Equal(t, []string{"ADM", "OPS"}, p.Requests("GET /api/experiments/schedules/v2")[0].Query["team"])
	assert.Contains(t, out, "0 0 9 ? * MON-FRI (Europe/Berlin)")
}

func TestGetLeavesOutWhatCannotBeSentBack(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("GET /api/experiments/schedules/"+id, platformtest.Reply{Body: `{"id":"` + id + `","experimentKey":"ADM-1","cron":"x","editedBy":{},"lastUpdated":"y","nextExecution":"z"}`})
	file := filepath.Join(t.TempDir(), "schedule.yml")

	_, err := platformtest.Stdout(t, func() error { return schedule.Get(ctx, p.Client, schedule.GetOptions{ID: id, File: file}) })

	require.NoError(t, err)
	content, _ := os.ReadFile(file)
	assert.Equal(t, "id: "+id+"\nexperimentKey: ADM-1\ncron: x\n", string(content))
}

func TestApplyCreatesAndWritesTheIdFirst(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/schedules", platformtest.Reply{Status: http.StatusCreated, JSON: saved})
	file := filepath.Join(t.TempDir(), "new.yml")
	require.NoError(t, os.WriteFile(file, []byte("experimentKey: ADM-1\nstartAt: 2030-06-01T09:00:00Z\n"), 0o644))

	out, err := platformtest.Stdout(t, func() error { return schedule.Apply(ctx, p.Client, schedule.ApplyOptions{Files: []string{file}}) })

	require.NoError(t, err)
	assert.Equal(t, "Experiment schedule "+id+" for ADM-1 created.\n", out)
	// A YAML timestamp was a JavaScript Date: sent, and written back, as an ISO string.
	assert.Equal(t, map[string]any{"experimentKey": "ADM-1", "startAt": "2030-06-01T09:00:00.000Z"}, p.Requests("POST /api/experiments/schedules")[0].JSON(t))
	content, _ := os.ReadFile(file)
	assert.Equal(t, "id: "+id+"\nexperimentKey: ADM-1\nstartAt: 2030-06-01T09:00:00.000Z\n", string(content))
}

func TestApplyKeepsAFileThatHasAnId(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/schedules", platformtest.Reply{JSON: saved})
	file := filepath.Join(t.TempDir(), "s.json")
	original := `{"id":"` + id + `","experimentKey":"ADM-1","editedBy":{"username":"jane"}}`
	require.NoError(t, os.WriteFile(file, []byte(original), 0o644))

	_, err := platformtest.Stdout(t, func() error { return schedule.Apply(ctx, p.Client, schedule.ApplyOptions{Files: []string{file}}) })

	require.NoError(t, err)
	assert.Equal(t, map[string]any{"id": id, "experimentKey": "ADM-1"}, p.Requests("POST /api/experiments/schedules")[0].JSON(t))
	content, _ := os.ReadFile(file)
	assert.Equal(t, original, string(content))
}

func TestCreateAndItsRules(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("POST /api/experiments/schedules", platformtest.Reply{Status: http.StatusCreated, JSON: saved})
	vars := jsyaml.NewMap()
	vars.Set("region", "eu")
	no := false

	_, err := platformtest.Stdout(t, func() error {
		return schedule.Create(ctx, p.Client, schedule.CreateOptions{Experiment: "ADM-1", Fields: schedule.Fields{Cron: "0 0 9 ? * *", Timezone: "Europe/Berlin", AllowParallel: &no, Variables: vars}})
	})

	require.NoError(t, err)
	assert.Equal(t, map[string]any{"experimentKey": "ADM-1", "cron": "0 0 9 ? * *", "timezone": "Europe/Berlin", "allowParallel": false, "variables": map[string]any{"region": "eu"}, "enabled": true},
		p.Requests("POST /api/experiments/schedules")[0].JSON(t))
	assert.EqualError(t, schedule.Create(ctx, p.Client, schedule.CreateOptions{Experiment: "ADM-1"}), "Either --cron or --start-at must be specified.")
	assert.EqualError(t, schedule.Create(ctx, p.Client, schedule.CreateOptions{Experiment: "ADM-1", Fields: schedule.Fields{Cron: "x", StartAt: "y"}}), "--cron and --start-at cannot be combined.")
}

func TestUpdateSendsOnlyWhatWasGiven(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("PATCH /api/experiments/schedules/"+id, platformtest.Reply{JSON: saved})

	out, err := platformtest.Stdout(t, func() error {
		if err := schedule.Update(ctx, p.Client, schedule.UpdateOptions{ID: id, Fields: schedule.Fields{Cron: "0 30 8 ? * *"}}); err != nil {
			return err
		}
		return schedule.SetEnabled(ctx, p.Client, id, false)
	})

	require.NoError(t, err)
	requests := p.Requests("PATCH /api/experiments/schedules/" + id)
	assert.Equal(t, map[string]any{"cron": "0 30 8 ? * *"}, requests[0].JSON(t))
	assert.Equal(t, map[string]any{"enabled": false}, requests[1].JSON(t))
	assert.Contains(t, out, "Experiment schedule "+id+" for ADM-1 disabled.")
	assert.EqualError(t, schedule.Update(ctx, p.Client, schedule.UpdateOptions{ID: id}), "Nothing to update. Pass at least one of the options, see --help.")
}

func TestDelete(t *testing.T) {
	p := platformtest.New(t)
	p.Reply("DELETE /api/experiments/schedules/"+id, platformtest.Reply{})
	p.Reply("DELETE /api/experiments/schedules/nope", platformtest.Reply{Status: http.StatusNotFound})

	out, err := platformtest.Stdout(t, func() error { return schedule.Delete(ctx, p.Client, id) })

	require.NoError(t, err)
	assert.Equal(t, "Experiment schedule "+id+" deleted.\n", out)
	assert.EqualError(t, schedule.Delete(ctx, p.Client, "nope"), "Experiment schedule nope not found.")
}

File: src/schedule/schedule.test.ts
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 Steadybit GmbH

import fs from 'node:fs/promises';
import { describe, expect, it, vi } from 'vitest';
import { respondTo } from '../mocks/recorder.ts';
import { writeFile } from '../mocks/tempFiles.ts';
import { load } from '../yaml.ts';
import {
  applySchedules,
  createSchedule,
  deleteSchedule,
  disableSchedule,
  enableSchedule,
  getSchedule,
  listSchedules,
  updateSchedule,
} from './commands.ts';

const ID = '01951394-727f-76a0-8675-c7519ebd0ff5';
const SCHEDULE = {
  id: ID,
  experimentKey: 'ADM-1',
  cron: '0 0 9 ? * MON-FRI',
  timezone: 'Europe/Berlin',
  enabled: true,
  allowParallel: false,
  editedBy: { username: 'jane' },
  lastUpdated: '2026-09-01T10:00:00Z',
};

describe('schedule', () => {
  describe('list', () => {
    it('filters by teams and experiments with repeated parameters', async () => {
      const requests = respondTo('get', '/api/experiments/schedules/v2', () => ({ json: [SCHEDULE] }));
      const logSpy = vi.spyOn(console, 'log');

      await listSchedules({ team: ['ADM', 'OPS'], experiment: ['ADM-1'] });

      expect(requests[0].url.searchParams.getAll('team')).toEqual(['ADM', 'OPS']);
      expect(requests[0].url.searchParams.getAll('experiment')).toEqual(['ADM-1']);
      const printed = logSpy.mock.calls.flat().join('\n');
      expect(printed).toContain(ID);
      expect(printed).toContain('0 0 9 ? * MON-FRI (Europe/Berlin)');
    });

    it('sends no filter when none is given', async () => {
      const requests = respondTo('get', '/api/experiments/schedules/v2', () => ({ json: [] }));
      const logSpy = vi.spyOn(console, 'log');

      await listSchedules({});

      expect(requests[0].url.search).toBe('');
      expect(logSpy).toHaveBeenCalledWith('No experiment schedules found.');
    });
  });

  describe('get', () => {
    it('leaves out the read-only fields so the file can be applied again', async () => {
      respondTo('get', `/api/experiments/schedules/${ID}`, () => ({ json: SCHEDULE }));
      const file = await writeFile('schedule.yml', {});

      await getSchedule({ id: ID, file });

      const expected: Partial<typeof SCHEDULE> = { ...SCHEDULE };
      delete expected.editedBy;
      delete expected.lastUpdated;
      expect(load(await fs.readFile(file, 'utf8'))).toEqual(expected);
    });

    it('reports a schedule that does not exist', async () => {
      respondTo('get', '/api/experiments/schedules/nope', () => ({ status: 404 }));

      await expect(getSchedule({ id: 'nope' })).rejects.toThrow('Experiment schedule nope not found.');
    });
  });

  describe('apply', () => {
    it('creates a schedule and writes its id back into the file', async () => {
      const requests = respondTo('post', '/api/experiments/schedules', () => ({ status: 201, json: SCHEDULE }));
      const file = await writeFile('new-schedule.yml', { experimentKey: 'ADM-1', cron: '0 0 9 ? * MON-FRI' });
      const logSpy = vi.spyOn(console, 'log');

      await applySchedules({ file: [file], recursive: false });

      expect(requests[0].body).toEqual({ experimentKey: 'ADM-1', cron: '0 0 9 ? * MON-FRI' });
      expect(load(await fs.readFile(file, 'utf8'))).toEqual({
        id: ID,
        experimentKey: 'ADM-1',
        cron: '0 0 9 ? * MON-FRI',
      });
      expect(logSpy).toHaveBeenCalledWith('Experiment schedule %s for %s %s.', ID, 'ADM-1', 'created');
    });

    it('updates a schedule that has an id without touching the file', async () => {
      const requests = respondTo('post', '/api/experiments/schedules', () => ({ status: 200, json: SCHEDULE }));
      const file = await writeFile('schedule.json', SCHEDULE, 'json');
      const before = await fs.readFile(file, 'utf8');

      await applySchedules({ file: [file], recursive: false });

      expect(requests[0].body).not.toHaveProperty('editedBy');
      expect(requests[0].body).toHaveProperty('id', ID);
      expect(await fs.readFile(file, 'utf8')).toBe(before);
    });

    it('rejects a file without an experiment key', async () => {
      const file = await writeFile('broken.yml', { cron: '* * * ? * *' });

      await expect(applySchedules({ file: [file], recursive: false })).rejects.toThrow(
        'does not name an experimentKey'
      );
    });
  });

  describe('create', () => {
    it('creates a schedule from flags', async () => {
      const requests = respondTo('post', '/api/experiments/schedules', () => ({ status: 201, json: SCHEDULE }));

      await createSchedule({
        experiment: 'ADM-1',
        cron: '0 0 9 ? * MON-FRI',
        timezone: 'Europe/Berlin',
        allowParallel: false,
        variable: { region: 'eu' },
      });

      expect(requests[0].body).toEqual({
        experimentKey: 'ADM-1',
        cron: '0 0 9 ? * MON-FRI',
        timezone: 'Europe/Berlin',
        allowParallel: false,
        enabled: true,
        variables: { region: 'eu' },
      });
    });

    it('creates a disabled one-off schedule', async () => {
      const requests = respondTo('post', '/api/experiments/schedules', () => ({ status: 201, json: SCHEDULE }));

      await createSchedule({ experiment: 'ADM-1', startAt: '2026-10-01T09:00:00Z', disabled: true });

      expect(requests[0].body).toEqual({ experimentKey: 'ADM-1', startAt: '2026-10-01T09:00:00Z', enabled: false });
    });

    it('needs either --cron or --start-at, not both', async () => {
      await expect(createSchedule({ experiment: 'ADM-1' })).rejects.toThrow(
        'Either --cron or --start-at must be specified.'
      );
      await expect(createSchedule({ experiment: 'ADM-1', cron: 'x', startAt: 'y' })).rejects.toThrow(
        '--cron and --start-at cannot be combined.'
      );
    });
  });

  describe('update', () => {
    it('patches only the given fields', async () => {
      const requests = respondTo('patch', `/api/experiments/schedules/${ID}`, () => ({ json: SCHEDULE }));

      await updateSchedule({ id: ID, cron: '0 30 8 ? * *' });

      expect(requests[0].body).toEqual({ cron: '0 30 8 ? * *' });
    });

    it('refuses an update without changes', async () => {
      await expect(updateSchedule({ id: ID })).rejects.toThrow('Nothing to update.');
    });

    it('enables and disables', async () => {
      const requests = respondTo('patch', `/api/experiments/schedules/${ID}`, () => ({ json: SCHEDULE }));
      const logSpy = vi.spyOn(console, 'log');

      await disableSchedule({ id: ID });
      await enableSchedule({ id: ID });

      expect(requests.map(r => r.body)).toEqual([{ enabled: false }, { enabled: true }]);
      expect(logSpy).toHaveBeenCalledWith('Experiment schedule %s for %s %s.', ID, 'ADM-1', 'disabled');
    });
  });

  describe('delete', () => {
    it('deletes a schedule', async () => {
      const requests = respondTo('delete', `/api/experiments/schedules/${ID}`, () => ({}));
      const logSpy = vi.spyOn(console, 'log');

      await deleteSchedule({ id: ID });

      expect(requests).toHaveLength(1);
      expect(logSpy).toHaveBeenCalledWith('Experiment schedule %s deleted.', ID);
    });

    it('reports a schedule that does not exist', async () => {
      respondTo('delete', '/api/experiments/schedules/nope', () => ({ status: 404 }));

      await expect(deleteSchedule({ id: 'nope' })).rejects.toThrow('Experiment schedule nope not found.');
    });
  });
});
